Per device listening

bentiss opened the pull request:
This goes on top of #10.

Implementation of the per-device listen() mechanism with the `StartDiscovery()` internal call.

We internally discussed about using `StopDiscovery()` to clean up the `RSSI` property, but this only works if we are the only one controlling the discovery.
However, we can remove the `DuplicateData` filter which will provide us all the `RSSI` events even if they are the same than the previous ones.

There are still a few bits left:
 - [ ] there is not `ListeningStopped` signal emitted [Update: do we need one?]
 - [x] ~~if `bluetoothd.service` is restarted, the bluez devices do not have their `ManufacturerData` set, so the devices are not created~~
 - [x] ~~if `bluetoothd.service` is restarted, and `StartDiscovery()` is called by an other mean, the communication fail with the new device~~
 - [ ] probably something else

bentiss committed:
README: document the new listening process

We need to be able to selectively enable/disable listening mode per
device, and the Listening property needs to be updated.

The timeout is there for the oinly reason we do not track client
requests. When we do so, we should disable the listening based on a
ref count.

bentiss committed:
implement a basic Start/StopListening

There is still no timeout, and no signal gets emitted.
The current way of testing this is:
- call StartListening() on the DBus device
- start discovery on the adapter by some other mean
- press the button on the device -> the sync will happen
- call StopDiscovery()
- press the button on the device -> the sync will *not* happen

bentiss committed:
dbus: keep track of the senders of the StartListening() method

If the sender disappear, we should stop listening for incoming events.

We match the Start/StopListening() that way, but if a client forgets
to call StopListening() before leaving and some data are being retrieved,
it's not our problem.

bentiss committed:
Start discovery mode when one device requests it

We need to check when the discovery stops (timeout from StartSearch)
if we should keep the discover one or not.

whot commented on README.md:
```diff
@@ -112,9 +112,25 @@ org.freedesktop.tuhi1.Device
+      property updates to the number of available drawings.
+      When the button is pressed multiple times, any new data is appended
+      to the existing list of drawings as long as this property is True.
+
```
We have a `Listening` property but also a `StoppedListening` signal - do we need both? It seems superfluous and while there's plenty of debugging use-cases where having the property is useful, the primary use-case for this is still one client doing everything, so the signal is enough. And the signal is what can carry the error code, so we do need that.

bentiss replied:
IIRC, I wanted to keep `Listening` for the use case where you have 2 clients, and it's a way of notifying the other client that the first one started listening on the device.

OTOH, the `StoppedListening` signal is meant for the client that requested the `StartListening`.
However, we should also target the `StoppedListening` signal to the client, because if one client has a grab on the device (see #11), its signal should be different than the other clients who tried to steal the grab.

whot replied:
Not sure how much effort it would be but we could still internally separate the various timeouts so you get the `ListeningStopped` signal even while the device itself is still listening and handling events for the other client.

bentiss replied:
Quick answer while trying to focus on sth else: the timeouts are not required anymore. In this branch we are tracking clients, so if a client disappears, we call `StopListening()` for it and this makes the whole timeouts moot

whot commented on README.md:
```diff
@@ -124,29 +140,30 @@ org.freedesktop.tuhi1.Device
-      memory, whichever happens earlier.  Use GetJSONData() to retrieve the
-      data from the daemon.
+      A client can not call twice StartListening() on the same device without
+      a matching StopListening() in the middle. This will result in a failure.
```
Is there a reason why we can't just let the client call `StartListening()` and ignore it once we're in progress? Same with `StopListening()` can just be a noop if we're not doing so already.

If we do want this to be a single call, we should return an error here instead of `()`


fwiw, I have a local FIXME commit here that fixes some of the grammar, will get to that tomorrow

bentiss replied:
> Is there a reason why we can't just let the client call `StartListening()` and ignore it once we're in progress? Same with `StopListening()` can just be a noop if we're not doing so already.

I mostly copied/pasted the behaviour of the bluez stack regarding `Start/StopDiscovery`.

The question is actually more whther or not we want to enforce the clients to behave sanely (matching start and stop) or we will just be happy having clients repeatively call `StartListening` because they didn't read the spec.

> If we do want this to be a single call, we should return an error here instead of ()

Ack to that

> fwiw, I have a local FIXME commit here that fixes some of the grammar, will get to that tomorrow

cool :)

whot commented on tuhi/dbusserver.py:
```diff
@@ -177,13 +177,47 @@ def _on_device_paired(self, device, pspec):
+        connection.signal_unsubscribe(s)
+        del(self._listening_clients[client])
+        logger.debug('Listening stopped on {} for {}'.format(self.name, client))
+
```
ACK to this commit

bentiss committed:
tuhi: handle cold-plugged devices

When bluez restarts (or the tuhi daemon restarts), the values we have
in the bluez device's field ManufacturerData are quite not accurate.

When bluez restarts they are empty, and if the last time we saw the
device was for the pairing process, the device will still be marked
as in the pairing mode.

So we should mark the cold-plug sequence differently from the hot-plug
one, and we should be more confident in the current configuration we
have stored to export the currently known devices over dbus.

Fixes #13

whot commented on tuhi.py:
```diff
@@ -264,8 +268,12 @@ def _on_bluez_device_updated(self, manager, bluez_device):
         if Tuhi._is_pairing_device(bluez_device):
             logger.debug('{}: call Pair() on device'.format(bluez_device.objpath))
+        elif d.listening:
+            d.connect_device()
```
maybe better to move the hunks from this file into the next commit

whot commented on tuhi/dbusserver.py:
```diff
@@ -217,7 +217,7 @@ def _stop_listening(self, connection, client):
 
-        # FIXME: notify the server to stop discovery
+        self.notify('listening')
 
```
ACK to this commit

whot reviewed (changes requested):
If you don't get to it, I'll pick it up tomorrow. The question regarding signal vs property is the main one here.

whot commented on tuhi.py:
```diff
@@ -253,18 +253,28 @@ def _on_bluez_discovery_stopped(self, manager):
 
-    def _on_bluez_device_updated(self, manager, bluez_device):
-        if bluez_device.vendor_id != WACOM_COMPANY_ID:
+    def _on_bluez_device_updated(self, manager, bluez_device, event=True):
```
API-wise, I'd really prefer the `event` parameter to be a `type` parameter and an enum. Much more readable having things like `if type == PRESENT_ON_START` or something than the quite ambiguous term "event"

bentiss replied:
That's what reviews are for. I'll try to work on it today.

whot reviewed (changes requested):
commit message: "quite not" -> "not quite"

whot commented:
superseded by #15
